src/server/pfs/server/commit_set.go

```diff
@@ -0,0 +1,455 @@
+	}
+	result := make([]*pfs.CommitInfo, 0)
+	collected := make(map[string]struct{})
+	collectCount := 0
+	// O(n^2) sorting of commits
+	//
+	// COULD WE instead sort by created time? DOES CREATED TIME COMPLETELY IMPLY TOPOLOGICAL ORDERING?
+	for collectCount < totalCommits {
+		for i, ci := range commitInfos {
+			satisfied := true
+			if _, ok := collected[pfsdb.RepoKey(ci.Commit.Repo)]; !ok {
+				for _, p := range ci.DirectProvenance {
+					if _, ok := collected[pfsdb.RepoKey(p.Repo)]; !ok {
+						satisfied = false
+						break
+					}
+				}
+			}
+			if satisfied {
+				collectCount++
+				collected[pfsdb.RepoKey(ci.Commit.Repo)] = struct{}{}
+				if !filterPassiveCommits || ci.Commit.ID == commitset.ID {
+					result = append(result, ci)
+				}
+				commitInfos[i] = commitInfos[len(commitInfos)-1]
+				commitInfos = commitInfos[:len(commitInfos)-1]
+				break
+			}
+		}
+	}
+	return result, nil
+}
+
+func (d *driver) inspectCommitSetImmediate(ctx context.Context, commitset *pfs.CommitSet, cb func(*pfs.CommitInfo) error) error {
+	var commitInfos []*pfs.CommitInfo
+	if err := d.txnEnv.WithReadContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
+		var err error
+		commitInfos, err = d.inspectCommitSetImmediateTx(txnCtx, commitset, false)
+		return err
+	}); err != nil {
+		return err
+	}
+	for _, commitInfo := range commitInfos {
+		if err := cb(commitInfo); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// applies a callback, cb, to all the commits in a commit set. A commit set includes
+func (d *driver) inspectCommitSet(ctx context.Context, commitset *pfs.CommitSet, wait bool, cb func(*pfs.CommitInfo) error) error {
+	if !wait {
+		return d.inspectCommitSetImmediate(ctx, commitset, cb)
+	}
+	sent := map[string]struct{}{}
+	send := func(ci *pfs.CommitInfo) error {
+		if _, ok := sent[pfsdb.CommitKey(ci.Commit)]; ok {
+			return nil
+		}
+		sent[pfsdb.CommitKey(ci.Commit)] = struct{}{}
+		return cb(ci)
+
+	}
+	unfinishedCommits := make([]*pfs.Commit, 0)
+	// NOTE: before 2.5, a triggered commits would be included as part of the same commit set as the triggering commit set,
+	// so we would wait for all of the current commit set to finish, then check if new previously unknown commits are added due to triggers.
+	if err := d.inspectCommitSetImmediate(ctx, commitset, func(ci *pfs.CommitInfo) error {
+		if ci.Finished != nil {
+			return send(ci)
+		}
+		unfinishedCommits = append(unfinishedCommits, proto.Clone(ci.Commit).(*pfs.Commit))
+		return nil
+	}); err != nil {
+		return err
+	}
+	for _, uc := range unfinishedCommits {
+		// TODO: make a dedicated call just for the blocking part, inspectCommit is a little heavyweight?
+		ci, err := d.inspectCommit(ctx, uc, pfs.CommitState_FINISHED)
+		if err != nil {
+			return err
+		}
+		if err := send(ci); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *driver) listCommitSet(ctx context.Context, project *pfs.Project, cb func(*pfs.CommitSetInfo) error) error {
```

Comment:
We are probably going to need a lightweight version of this. Resolving all of the commits for all of the commit sets is going to be costly in many workloads. Even with caching, this is probably moving a lot more data than the user cares about in most cases.

Reply:
FWIW, the only use of this right now is the CLI